Infinite loading game bar when leaving a remote game

Noticed this while testing the other chat issues.
Steps to reproduce:
Join a bot
Start the game
Press X to close the window or press File > Leave Game
Get a loading dialog that won't close what so ever.

If you press File > Exit Game everything works fine.

It is still occurring with the latest as well as this error @RoiEXLab 
When attempting to use toolbox to view a bot host.

TripleA engine version 1.9.0.0.10355
Loading map: , from: C:\Users\prast\triplea\downloadedMaps\
Loading resources from the following paths: [C:\Users\prast\triplea\downloadedMaps\, G:\10355\TripleA\assets]
Error: Exception on remote
java.lang.RuntimeException: Exception on remote
	at games.strategy.engine.message.UnifiedInvocationHandler.invoke(UnifiedInvocationHandler.java:54)
	at com.sun.proxy.$Proxy2.getHostConnections(Unknown Source)
	at games.strategy.engine.lobby.client.ui.LobbyGamePanel.getHostInfo(LobbyGamePanel.java:275)
	at games.strategy.engine.lobby.client.ui.LobbyGamePanel.lambda$getGeneralAdminGamesListContextActions$6(LobbyGamePanel.java:196)
	at games.strategy.ui.SwingAction$1.actionPerformed(SwingAction.java:65)
	at javax.swing.AbstractButton.fireActionPerformed(Unknown Source)
	at javax.swing.AbstractButton$Handler.actionPerformed(Unknown Source)
	at javax.swing.DefaultButtonModel.fireActionPerformed(Unknown Source)
	at javax.swing.DefaultButtonModel.setPressed(Unknown Source)
	at javax.swing.AbstractButton.doClick(Unknown Source)
	at javax.swing.plaf.basic.BasicMenuItemUI.doClick(Unknown Source)
	at javax.swing.plaf.basic.BasicMenuItemUI$Handler.mouseReleased(Unknown Source)
	at java.awt.AWTEventMulticaster.mouseReleased(Unknown Source)
	at java.awt.Component.processMouseEvent(Unknown Source)
	at javax.swing.JComponent.processMouseEvent(Unknown Source)
	at java.awt.Component.processEvent(Unknown Source)
	at java.awt.Container.processEvent(Unknown Source)
	at java.awt.Component.dispatchEventImpl(Unknown Source)
	at java.awt.Container.dispatchEventImpl(Unknown Source)
	at java.awt.Component.dispatchEvent(Unknown Source)
	at java.awt.LightweightDispatcher.retargetMouseEvent(Unknown Source)
	at java.awt.LightweightDispatcher.processMouseEvent(Unknown Source)
	at java.awt.LightweightDispatcher.dispatchEvent(Unknown Source)
	at java.awt.Container.dispatchEventImpl(Unknown Source)
	at java.awt.Window.dispatchEventImpl(Unknown Source)
	at java.awt.Component.dispatchEvent(Unknown Source)
	at java.awt.EventQueue.dispatchEventImpl(Unknown Source)
	at java.awt.EventQueue.access$500(Unknown Source)
	at java.awt.EventQueue$3.run(Unknown Source)
	at java.awt.EventQueue$3.run(Unknown Source)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(Unknown Source)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(Unknown Source)
	at java.awt.EventQueue$4.run(Unknown Source)
	at java.awt.EventQueue$4.run(Unknown Source)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(Unknown Source)
	at java.awt.EventQueue.dispatchEvent(Unknown Source)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(Unknown Source)
	at java.awt.EventDispatchThread.pumpEventsForFilter(Unknown Source)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(Unknown Source)
	at java.awt.EventDispatchThread.pumpEvents(Unknown Source)
	at java.awt.EventDispatchThread.pumpEvents(Unknown Source)
	at java.awt.EventDispatchThread.run(Unknown Source)
Caused by: games.strategy.engine.message.RemoteNotFoundException: Not found:games.strategy.engine.lobby.server.RemoteHostUtils:Bot201_CA_USA_lobby_watcher (1)_lobby_watcher port:59574 ip:173.255.243.147
	at games.strategy.engine.message.UnifiedMessengerHub.messageReceived(UnifiedMessengerHub.java:89)
	at games.strategy.engine.message.unifiedmessenger.UnifiedMessenger.send(UnifiedMessenger.java:220)
	at games.strategy.engine.message.unifiedmessenger.UnifiedMessenger.invokeAndWaitRemote(UnifiedMessenger.java:123)
	at games.strategy.engine.message.unifiedmessenger.UnifiedMessenger.invokeAndWait(UnifiedMessenger.java:100)
	at games.strategy.engine.message.UnifiedInvocationHandler.invoke(UnifiedInvocationHandler.java:52)
	at com.sun.proxy.$Proxy4.getConnections(Unknown Source)
	at games.strategy.engine.lobby.server.ModeratorController.getHostConnections(ModeratorController.java:352)
	at sun.reflect.GeneratedMethodAccessor35.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at games.strategy.engine.message.unifiedmessenger.EndPoint.invokeSingle(EndPoint.java:157)
	at games.strategy.engine.message.unifiedmessenger.EndPoint.invokeMultiple(EndPoint.java:140)
	at games.strategy.engine.message.unifiedmessenger.EndPoint.invokeLocal(EndPoint.java:126)
	at games.strategy.engine.message.unifiedmessenger.UnifiedMessenger.lambda$messageReceived$1(UnifiedMessenger.java:275)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.IllegalStateException: remote not found
	at java.lang.Thread.getStackTrace(Thread.java:1559)
	at games.strategy.engine.message.MessengerException.fillInInvokerStackTrace(MessengerException.java:19)
	at games.strategy.engine.message.UnifiedInvocationHandler.invoke(UnifiedInvocationHandler.java:56)
	... 12 more
Error: Exception on remote
java.lang.RuntimeException: Exception on remote
	at games.strategy.engine.message.UnifiedInvocationHandler.invoke(UnifiedInvocationHandler.java:54)
	at com.sun.proxy.$Proxy2.getHostConnections(Unknown Source)
	at games.strategy.engine.lobby.client.ui.LobbyGamePanel.getHostInfo(LobbyGamePanel.java:275)
	at games.strategy.engine.lobby.client.ui.LobbyGamePanel.lambda$getGeneralAdminGamesListContextActions$6(LobbyGamePanel.java:196)
	at games.strategy.ui.SwingAction$1.actionPerformed(SwingAction.java:65)
	at javax.swing.AbstractButton.fireActionPerformed(Unknown Source)
	at javax.swing.AbstractButton$Handler.actionPerformed(Unknown Source)
	at javax.swing.DefaultButtonModel.fireActionPerformed(Unknown Source)
	at javax.swing.DefaultButtonModel.setPressed(Unknown Source)
	at javax.swing.AbstractButton.doClick(Unknown Source)
	at javax.swing.plaf.basic.BasicMenuItemUI.doClick(Unknown Source)
	at javax.swing.plaf.basic.BasicMenuItemUI$Handler.mouseReleased(Unknown Source)
	at java.awt.AWTEventMulticaster.mouseReleased(Unknown Source)
	at java.awt.Component.processMouseEvent(Unknown Source)
	at javax.swing.JComponent.processMouseEvent(Unknown Source)
	at java.awt.Component.processEvent(Unknown Source)
	at java.awt.Container.processEvent(Unknown Source)
	at java.awt.Component.dispatchEventImpl(Unknown Source)
	at java.awt.Container.dispatchEventImpl(Unknown Source)
	at java.awt.Component.dispatchEvent(Unknown Source)
	at java.awt.LightweightDispatcher.retargetMouseEvent(Unknown Source)
	at java.awt.LightweightDispatcher.processMouseEvent(Unknown Source)
	at java.awt.LightweightDispatcher.dispatchEvent(Unknown Source)
	at java.awt.Container.dispatchEventImpl(Unknown Source)
	at java.awt.Window.dispatchEventImpl(Unknown Source)
	at java.awt.Component.dispatchEvent(Unknown Source)
	at java.awt.EventQueue.dispatchEventImpl(Unknown Source)
	at java.awt.EventQueue.access$500(Unknown Source)
	at java.awt.EventQueue$3.run(Unknown Source)
	at java.awt.EventQueue$3.run(Unknown Source)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(Unknown Source)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(Unknown Source)
	at java.awt.EventQueue$4.run(Unknown Source)
	at java.awt.EventQueue$4.run(Unknown Source)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(Unknown Source)
	at java.awt.EventQueue.dispatchEvent(Unknown Source)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(Unknown Source)
	at java.awt.EventDispatchThread.pumpEventsForFilter(Unknown Source)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(Unknown Source)
	at java.awt.EventDispatchThread.pumpEvents(Unknown Source)
	at java.awt.EventDispatchThread.pumpEvents(Unknown Source)
	at java.awt.EventDispatchThread.run(Unknown Source)
Caused by: games.strategy.engine.message.RemoteNotFoundException: Not found:games.strategy.engine.lobby.server.RemoteHostUtils:Bot201_CA_USA_lobby_watcher (1)_lobby_watcher port:59574 ip:173.255.243.147
	at games.strategy.engine.message.UnifiedMessengerHub.messageReceived(UnifiedMessengerHub.java:89)
	at games.strategy.engine.message.unifiedmessenger.UnifiedMessenger.send(UnifiedMessenger.java:220)
	at games.strategy.engine.message.unifiedmessenger.UnifiedMessenger.invokeAndWaitRemote(UnifiedMessenger.java:123)
	at games.strategy.engine.message.unifiedmessenger.UnifiedMessenger.invokeAndWait(UnifiedMessenger.java:100)
	at games.strategy.engine.message.UnifiedInvocationHandler.invoke(UnifiedInvocationHandler.java:52)
	at com.sun.proxy.$Proxy4.getConnections(Unknown Source)
	at games.strategy.engine.lobby.server.ModeratorController.getHostConnections(ModeratorController.java:352)
	at sun.reflect.GeneratedMethodAccessor35.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at games.strategy.engine.message.unifiedmessenger.EndPoint.invokeSingle(EndPoint.java:157)
	at games.strategy.engine.message.unifiedmessenger.EndPoint.invokeMultiple(EndPoint.java:140)
	at games.strategy.engine.message.unifiedmessenger.EndPoint.invokeLocal(EndPoint.java:126)
	at games.strategy.engine.message.unifiedmessenger.UnifiedMessenger.lambda$messageReceived$1(UnifiedMessenger.java:275)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.IllegalStateException: remote not found
	at java.lang.Thread.getStackTrace(Thread.java:1559)
	at games.strategy.engine.message.MessengerException.fillInInvokerStackTrace(MessengerException.java:19)
	at games.strategy.engine.message.UnifiedInvocationHandler.invoke(UnifiedInvocationHandler.java:56)
	... 12 more

--- a/game-core/src/main/java/games/strategy/engine/framework/GameRunner.java
+++ b/game-core/src/main/java/games/strategy/engine/framework/GameRunner.java
@@ -137,6 +137,19 @@ public class GameRunner {
       return;
     }
 
+    if (SystemProperties.isMac()) {
+      com.apple.eawt.Application.getApplication().setOpenURIHandler(event -> {
+        final String encoding = StandardCharsets.UTF_8.displayName();
+        try {
+          final String mapName = URLDecoder.decode(
+              event.getURI().toString().substring(ArgParser.TRIPLEA_PROTOCOL.length()), encoding);
+          SwingUtilities.invokeLater(() -> DownloadMapsWindow.showDownloadMapsWindowAndDownload(mapName));
+        } catch (final UnsupportedEncodingException e) {
+          throw new AssertionError(encoding + " is not a supported encoding!", e);
+        }
+      });
+    }
+
     if (HttpProxy.isUsingSystemProxy()) {
       HttpProxy.updateSystemProxy();
     }
@@ -274,20 +287,6 @@ public class GameRunner {
    * welcome (launch lobby/single player game etc..) screen presented to GUI enabled clients.
    */
   public static void showMainFrame() {
-
-    if (SystemProperties.isMac()) {
-      com.apple.eawt.Application.getApplication().setOpenURIHandler(event -> {
-        final String encoding = StandardCharsets.UTF_8.displayName();
-        try {
-          final String mapName = URLDecoder.decode(
-              event.getURI().toString().substring(ArgParser.TRIPLEA_PROTOCOL.length()), encoding);
-          SwingUtilities.invokeLater(() -> DownloadMapsWindow.showDownloadMapsWindowAndDownload(mapName));
-        } catch (final UnsupportedEncodingException e) {
-          throw new AssertionError(encoding + " is not a supported encoding!", e);
-        }
-      });
-    }
-
     SwingUtilities.invokeLater(() -> {
       mainFrame.requestFocus();
       mainFrame.toFront();

--- a/game-core/src/main/java/games/strategy/engine/framework/GameRunner.java
+++ b/game-core/src/main/java/games/strategy/engine/framework/GameRunner.java
@@ -451,13 +451,14 @@ public class GameRunner {
    * After the game has been left, call this.
    */
   public static void clientLeftGame() {
-    Interruptibles.await(() -> SwingAction.invokeAndWait(() -> {
-      // having an oddball issue with the zip stream being closed while parsing to load default game. might be caused by
-      // closing of stream while unloading map resources.
-      Interruptibles.sleep(100);
-      setupPanelModel.showSelectType();
-      showMainFrame();
-    }));
+    if (SwingUtilities.isEventDispatchThread()) {
+      throw new IllegalStateException("This method must not be called from the EDT");
+    }
+    // having an oddball issue with the zip stream being closed while parsing to load default game. might be caused by
+    // closing of stream while unloading map resources.
+    Interruptibles.sleep(100);
+    Interruptibles.await(() -> SwingAction.invokeAndWait(setupPanelModel::showSelectType));
+    showMainFrame();
   }
 
   public static void quitGame() {

--- a/game-core/src/main/java/games/strategy/engine/framework/GameRunner.java
+++ b/game-core/src/main/java/games/strategy/engine/framework/GameRunner.java
@@ -300,8 +300,10 @@ public class GameRunner {
 
       if (System.getProperty(TRIPLEA_SERVER, "false").equals("true")) {
         setupPanelModel.showServer(mainFrame);
+        System.clearProperty(TRIPLEA_SERVER);
       } else if (System.getProperty(TRIPLEA_CLIENT, "false").equals("true")) {
         setupPanelModel.showClient(mainFrame);
+        System.clearProperty(TRIPLEA_CLIENT);
       }
     });
   }
@@ -310,7 +312,7 @@ public class GameRunner {
     Interruptibles.await(() -> newBackgroundTaskRunner().runInBackground("Loading game...", () -> {
       gameSelectorModel.loadDefaultGameSameThread();
       final String fileName = System.getProperty(TRIPLEA_GAME, "");
-      if (fileName.length() > 0) {
+      if (!fileName.isEmpty()) {
         gameSelectorModel.load(new File(fileName), mainFrame);
       }
 

--- a/game-core/src/main/java/games/strategy/engine/framework/startup/mc/ClientModel.java
+++ b/game-core/src/main/java/games/strategy/engine/framework/startup/mc/ClientModel.java
@@ -153,6 +153,7 @@ public class ClientModel implements IMessengerErrorListener {
 
   public void setRemoteModelListener(@Nonnull final IRemoteModelListener listener) {
     this.listener = Preconditions.checkNotNull(listener);
+    internalPlayerListingChanged(getServerStartup().getPlayerListing());
   }
 
   private static ClientProps getProps(final Component ui) {
@@ -379,16 +380,15 @@ public class ClientModel implements IMessengerErrorListener {
   }
 
   private void internalPlayerListingChanged(final PlayerListing listing) {
-    SwingUtilities
-        .invokeLater(() -> gameSelectorModel.clearDataButKeepGameInfo(listing.getGameName(), listing.getGameRound(),
-            listing.getGameVersion().toString()));
+    gameSelectorModel.clearDataButKeepGameInfo(listing.getGameName(), listing.getGameRound(),
+            listing.getGameVersion().toString());
     synchronized (this) {
       playersToNodes = listing.getPlayerToNodeListing();
       playersEnabledListing = listing.getPlayersEnabledListing();
       playersAllowedToBeDisabled = listing.getPlayersAllowedToBeDisabled();
       playerNamesAndAlliancesInTurnOrder = listing.getPlayerNamesAndAlliancesInTurnOrderLinkedHashMap();
     }
-    SwingUtilities.invokeLater(() -> listener.playerListChanged());
+    listener.playerListChanged();
   }
 
   /**

--- a/game-core/src/main/java/games/strategy/engine/framework/startup/ui/ClientSetupPanel.java
+++ b/game-core/src/main/java/games/strategy/engine/framework/startup/ui/ClientSetupPanel.java
@@ -36,7 +36,7 @@ public class ClientSetupPanel extends SetupPanel {
   private static final long serialVersionUID = 6942605803526295372L;
   private final Insets buttonInsets = new Insets(0, 0, 0, 0);
   private final ClientModel clientModel;
-  private List<PlayerRow> playerRows = Collections.emptyList();
+  private final List<PlayerRow> playerRows = new ArrayList<>();
 
   public ClientSetupPanel(final ClientModel model) {
     clientModel = model;
@@ -48,7 +48,7 @@ public class ClientSetupPanel extends SetupPanel {
 
       @Override
       public void playerListChanged() {
-        internalPlayersChanged();
+        SwingUtilities.invokeLater(ClientSetupPanel.this::internalPlayersChanged);
       }
     });
   }
@@ -63,7 +63,7 @@ public class ClientSetupPanel extends SetupPanel {
       // clients only get to change bot settings
       disableable.clear();
     }
-    playerRows = new ArrayList<>();
+    playerRows.clear();
     final Set<String> playerNames = playerNamesAndAlliancesInTurnOrder.keySet();
     for (final String name : playerNames) {
       final PlayerRow playerRow = new PlayerRow(name, playerNamesAndAlliancesInTurnOrder.get(name),

--- a/game-core/src/main/java/games/strategy/triplea/ui/TripleAFrame.java
+++ b/game-core/src/main/java/games/strategy/triplea/ui/TripleAFrame.java
@@ -614,7 +614,7 @@ public class TripleAFrame extends MainGameFrame {
       ((ClientGame) game).shutDown();
       // an ugly hack, we need a better
       // way to get the main frame
-      GameRunner.clientLeftGame();
+      new Thread(GameRunner::clientLeftGame).start();
     }
   }
 

